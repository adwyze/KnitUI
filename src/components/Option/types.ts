import { CSSProperties, SyntheticEvent, ReactNode, FC } from "react"
import { IIconProps } from "../Icon/Icon"

export type TIconType = {
  left?: string | ReactNode
  right?: string | ReactNode
}

export interface IOptionProps {
  value: string | number
  disabled?: boolean
  customStyles?: CSSProperties
  onClick?: (e: SyntheticEvent, value: string | number) => void
}

export interface IOption extends FC<IOptionProps> {
  Tag: FC
  Avatar: FC
  Text: FC
  Icon: FC<IOptionIconProps>
  OptionGroup: FC<IOptionGroupProps>
}

interface OptionTextProps {
  description: string
  children: ReactNode
}

export interface IOptionText extends FC<OptionTextProps> {}

export interface IOptionGroupProps {
  title?: string
  children?: ReactNode[] | ReactNode
}

export interface IOptionIconProps extends IIconProps {
  align?: string
}
