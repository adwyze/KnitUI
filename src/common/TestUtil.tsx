import React, { ReactElement } from "react"
import { render } from "react-testing-library"
import { ThemeProvider } from "./styles"

const ProviderWrapper = ({
  children,
}: {
  children: ReactElement
}): ReactElement => {
  // Nest all providers here
  return <ThemeProvider>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: object) =>
  render(ui, { wrapper: ProviderWrapper, ...options })

// re-export everything
export * from "react-testing-library"
// override render method
export { customRender as render }
