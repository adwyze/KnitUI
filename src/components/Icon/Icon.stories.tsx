import React from "react"
import { storiesOf } from "@storybook/react"
import Icon from "./index"
import * as Icons from "./Icons"
import { withKnobs, select } from "@storybook/addon-knobs"

const stories = storiesOf("Icons", module)
stories.addDecorator(withKnobs)

const sizes = ["16px", "18px", "24px"]
const colors = ["#000000", "#990000", "#036600"]
const allIcons = Object.entries(Icons)

stories
  .add("Icon", () => {
    const size = select("Size", sizes, "18px")
    const color = select("Color", colors, "#000000")
    return <Icon type="oAddCircle" size={size} fill={color} />
  })
  .add("All icons", () => {
    const size = select("Size", sizes, "18px")
    const color = select("Color", colors, "#000000")
    const props = { width: size, height: size, fill: color }
    const allIconsDOM = allIcons.map(([iconKey, icon], index) => (
      <div
        key={index}
        title={iconKey}
        style={{ padding: 10, margin: 10 }}
        role="icon-wrapper">
        {icon(props)}
      </div>
    ))
    return (
      <div
        style={{
          width: "100%",
          display: "flex",
          flexWrap: "wrap",
        }}>
        {allIconsDOM}
      </div>
    )
  })
