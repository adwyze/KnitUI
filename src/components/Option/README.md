# Option

## Usage

```javascript
import { Option } from "KnitUI"

// render
return (
  <Option.Group title="SOME ITEMS">
    <Option onClick={(e, val) => alert(val)} value="Option 1">
      {/* Give an align prop to to set position, same API as Icon component otherwise */}
      <Option.Icon type="oClose" align="right" />
      {/* Add a description prop to text to add additional context */}
      <Option.Text description="Option Description">Option 1</Option.Text>
    </Option>
    <Option onClick={(e, val) => alert(val)} value="Option 2">
      {/* Same API as the Avatar component */}
      <Option.Avatar name="anshuman-kun" />
      <Option.Text description="Option Description">Option 2</Option.Text>
    </Option>
    <Option value="Option 3" selected>
      <Option.Text description="Option Description">Option 3</Option.Text>
      <Option.Tag text="Aa" align="right" />
    </Option>
  </Option.Group>
)

// If you don't pass any children to option then it will render Option.Text with the option value as text
```

# SubComponents

```js
// Same API as the original components
```

- Option.Icon
- Option.Avatar
- Option.Tag
- Option.Text
- Option.Group

### Option Props

| Prop name | Type           | Optional | Default | Description                                         |
| --------- | -------------- | -------- | ------- | --------------------------------------------------- |
| disabled  | boolean        | Yes      | `false` | Disable an option                                   |
| selected  | boolean        | Yes      | `false` | Mark an option as selected                          |
| onClick   | function       | Yes      | None    | Gives the event and option value in the parameters  |
| style     | CSSStyleObject | Yes      | None    | A style object to override default styled of option |
| className | string         | Yes      | None    | CSS class name to be added                          |

### Option.Group Props

| Prop name | Type      | Optional | Default                        | Description                |
| --------- | --------- | -------- | ------------------------------ | -------------------------- |
| title     | string    | Yes      | `` | Title of the option group |
| children  | ReactNode | Yes      | null                           | Options to render as child |

### Option.Text Props

| Prop name   | Type                | Optional | Default                               | Description                                           |
| ----------- | ------------------- | -------- | ------------------------------------- | ----------------------------------------------------- |
| description | string              | Yes      | `` | Add a description to option text |
| children    | string \| ReactNode | No       | null                                  | Option text or ReactNode to render as child of option |
