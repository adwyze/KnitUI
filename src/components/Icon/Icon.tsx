import React from "react"
import styled from "styled-components"
import * as Icons from "./Icons"
import { withProps } from "../../common/_utils"

export interface IIconProps {
  type: string
  height?: string
  width?: string
  fill?: string
  customStyles?: string
}

const StyledIconWrapper: any = withProps<IIconProps>()(styled.span)`
  cursor: inherit;
  display: inline-block;
  height: ${({ height }) => height};
  width: ${({ width }) => width};
  & svg {
   ${({ customStyles }) => customStyles};
  }
`

const Icon: any = (props: IIconProps) => {
  const {
    type,
    height = "18px",
    width = "18px",
    fill = "#000",
    customStyles = "",
  } = props
  const svgStyles = { height, width, fill }
  // If type is not provided, then "oCheckBoxOutlineBlank" is used as placeholder icon
  const defaultType = "oCheckBoxOutlineBlank"
  const icon =
    type && Icons[type] ? Icons[type](svgStyles) : Icons[defaultType](svgStyles)
  return (
    <StyledIconWrapper
      {...props}
      height={height}
      width={width}
      customStyles={customStyles}>
      {icon}
    </StyledIconWrapper>
  )
}

export default Icon
