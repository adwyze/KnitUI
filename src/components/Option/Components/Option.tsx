import React, { Children, cloneElement, ReactElement } from "react"
import Avatar from "../../Avatar"
import Icon from "../../Icon"
import Label from "../../Label"
import { IOption } from "../types"
import { FlexWrapper, OptionItemWrapper, OptionWrapper } from "./Option.styles"
import OptionGroup from "./OptionGroup"
import OptionText from "./OptionText"

const Option: IOption = props => {
  const {
    value,
    customStyles,
    onClick,
    children,
    disabled = false,
    ...rest
  } = props

  // Pass disabled props to all children
  const updatedChildren = Children.map(children, child => {
    return cloneElement(child as ReactElement, { disabled })
  })

  // Filter children based on their alignment
  const rightChildren = Children.toArray(children).filter(
    comp => comp!.props.align === "right"
  )
  const leftChildren = Children.toArray(children).filter(
    comp => comp!.props.align !== "right"
  )

  return (
    <OptionWrapper
      onClick={onClick ? e => onClick(e, value) : () => null}
      style={customStyles}
      isDisabled={disabled}
      {...rest}>
      <FlexWrapper>{leftChildren}</FlexWrapper>
      <FlexWrapper>{rightChildren}</FlexWrapper>
    </OptionWrapper>
  )
}

// Wrap all option sub-components in a common wrapper and pass all props to the corresponding component
Option.Icon = OptionItemWrapper(Icon)
Option.Text = OptionItemWrapper(OptionText)
Option.Tag = OptionItemWrapper(Label)
Option.Avatar = OptionItemWrapper(Avatar, { alignStart: true })
Option.OptionGroup = OptionGroup

export default Option
