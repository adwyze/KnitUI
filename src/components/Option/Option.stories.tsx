import React from "react"
import { withKnobs } from "@storybook/addon-knobs"
import { storiesOf } from "@storybook/react"
import Option from "./"
import Switch from "../Switch"
import Button from "../Button"
const Readme = require("./README.md")

storiesOf("Option", module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      // Show readme before story
      codeTheme: "shades-of-purple",
      // Show readme at the addons panel
      sidebar: Readme,
    },
  })
  .add("Basic option", () => (
    <Option onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Text>Option 1</Option.Text>
    </Option>
  ))
  .add("Option without child", () => (
    <Option onClick={(e, val) => alert(val)} value="Option 1" />
  ))
  .add("Disabled option", () => (
    <Option onClick={() => console.log("asd")} disabled value="Option 1">
      <Option.Text>Disabled Option</Option.Text>
    </Option>
  ))
  .add("With some icons", () => (
    <Option onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Icon type="oClose" align="right" />
      <Option.Icon type="oStorage" />
      <Option.Icon type="oCheckBox" align="right" />
      <Option.Text description="Option Description">Option 1</Option.Text>
    </Option>
  ))
  .add("With Option Group", () => (
    <Option.Group title="SOME ITEMS">
      <Option onClick={(e, val) => alert(val)} value="Option 1">
        <Option.Icon type="oClose" fill="yellow" align="right" />
        <Option.Icon type="oSettings" align="left" />
        <Option.Icon type="oClose" align="right" />
        <Option.Text description="Option Description">Option 1</Option.Text>
      </Option>
      <Option onClick={(e, val) => alert(val)} value="Option 2">
        <Option.Icon type="oClose" fill="yellow" align="right" />
        <Option.Icon type="oOpenWith" align="left" />
        <Option.Icon type="oClose" align="right" />
        <Option.Text description="Option Description">Option 2</Option.Text>
      </Option>
      <Option onClick={(e, val) => alert(val)} value="Option 3">
        <Option.Icon type="oClose" fill="yellow" align="right" />
        <Option.Icon type="oOpenWith" align="left" />
        <Option.Icon type="oClose" align="right" />
        <Option.Text description="Option Description">Option 3</Option.Text>
      </Option>
    </Option.Group>
  ))
  .add("With Selected Option", () => (
    <Option.Group title="SOME ITEMS">
      <Option onClick={(e, val) => alert(val)} value="Option 1">
        <Option.Text>Option 1</Option.Text>
      </Option>
      <Option selected onClick={(e, val) => alert(val)} value="Option 2">
        <Option.Text>Option 2</Option.Text>
      </Option>
      <Option onClick={(e, val) => alert(val)} value="Option 3">
        <Option.Text>Option 3</Option.Text>
      </Option>
    </Option.Group>
  ))
  .add("With Avatar", () => (
    <Option onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Avatar name="anshumankun" />
      <Option.Text description="Option Description">Option 1</Option.Text>
      <Option.Icon type="oClose" align="right" />
    </Option>
  ))
  .add("With disabled and Avatar", () => (
    <Option disabled onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Avatar name="anshumankun" />
      <Option.Text description="Option Description">Option 1</Option.Text>
      <Option.Icon type="oClose" align="right" />
    </Option>
  ))
  .add("With Tag and selected", () => (
    <Option selected onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Text>Option 1</Option.Text>
      <Option.Tag text="Aa" align="right" />
    </Option>
  ))
  .add("With selectors", () => (
    <Option.Group title="Selector">
      <Option value="Option 1">
        <Option.Text>Option 1</Option.Text>
        <Switch align="right" />
      </Option>
      <Option value="Option 2">
        <Option.Text>Option 2</Option.Text>
        <Switch align="right" />
      </Option>
    </Option.Group>
  ))
  .add("With overflow handling", () => (
    <div style={{ width: '20rem', border: '1px solid red'}}>
      <Option.Group title="Selector">
        <Option selected value="Option 1">
          <Option.Text>This is some really long option</Option.Text>
        </Option>
        <Option value="Option 2">
          <Option.Text>Option 2</Option.Text>
        </Option>
      </Option.Group>

    </div>
  ))
  .add("With custom fill", () => (
      <Option.Group title="Selector">
        <Option customStyles={{ background: '#FDEDCE' }} selected value="Option 1">
          <Option.Text>Campaign</Option.Text>
        </Option>
        <Option customStyles={{ background: '#DDEEE5' }} value="FB Spend">
          <Option.Text>FB Spend</Option.Text>
        </Option>
      </Option.Group>
  ))
  .add("With divider", () => (
    <Option.Group title="SOME ITEMS">
      <Option value="Edit">
        <Option.Icon type="oEdit" />
        <Option.Text>Edit</Option.Text>
      </Option><Option value="Duplicate">
        <Option.Icon type="oFileCopy" />
        <Option.Text>Duplicate</Option.Text>
      </Option><Option value="Move to">
        <Option.Icon type="oArrowForward" />
        <Option.Text>Move to</Option.Text>
      </Option>
      <Option.Divider />
      <Option value="Delete">
        <Option.Icon type="oDelete" />
        <Option.Text>Delete</Option.Text>
      </Option>
    </Option.Group>
  ))
  .add("With Button", () => (
    <Option onClick={(e, val) => alert(val)} value="Option 1">
      <Option.Text description="Option Description">Option 1</Option.Text>
      <Button onClick={(e) => e.stopPropagation()} align="right" label="asd">Click!</Button>
    </Option>
  ))
