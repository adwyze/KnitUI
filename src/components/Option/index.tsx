import Option from "./Components/Option"
export default Option
