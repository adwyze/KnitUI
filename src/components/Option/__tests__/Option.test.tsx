import React from "react"
import { cleanup, fireEvent, render } from "../../../common/TestUtil"
import Icon from "../../Icon"
import Option from ".."

afterEach(cleanup)

describe("Option", () => {
  it("renders basic Option correctly", () => {
    const { asFragment, container } = render(<Option value="Tanjiro" />)
    expect(container.children).toHaveLength(1)
    expect(container).toHaveTextContent("Tanjiro")
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders disabled option correctly", () => {
    const onClick = jest.fn()
    const { asFragment, container } = render(
      <Option disabled onClick={onClick} value="Tanjiro">
        <Option.Text description="this is the option description">
          Nezuko
        </Option.Text>
      </Option>
    )

    expect(container).toHaveTextContent("this is the option description")
    // Should not fire the onClick handler
    fireEvent.click(container)
    expect(onClick).not.toHaveBeenCalled()
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders selected option correctly and fires onClick", () => {
    const onClick = jest.fn()
    const { asFragment, container } = render(
      <Option selected value="Tanjiro">
        <Option.Text>Nezuko</Option.Text>
      </Option>
    )
    // should fire the given onClick
    fireEvent.click(container)
    expect(onClick).not.toHaveBeenCalled()
    // should have a check icon if selected props is supplied
    const icon = container.querySelectorAll('[type="oCheck"]')
    expect(icon).toHaveLength(1)
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders OptionGroup Correctly", () => {
    const { asFragment, container } = render(
      <Option.OptionGroup title="SOME ITEMS">
        <Option value="Tanjiro">
          <Option.Text description="Kimetsu no Yaiba">Option 1</Option.Text>
        </Option>
        <Option value="Zenitsu">
          <Option.Text description="Kimetsu no Yaiba">Option 2</Option.Text>
        </Option>
      </Option.OptionGroup>
    )
    // Should print the option group title
    expect(container).toHaveTextContent("SOME ITEMS")
    // should render the group title with children
    expect(container.children[0].children).toHaveLength(3)
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders icons and avatar with correct alignment ", () => {
    const { asFragment, getByTestId, container } = render(
      <Option value="Tanjiro">
        <Option.Avatar name="anshumankun" />
        <Option.Text description="Kimetsu no Yaiba">Option 1</Option.Text>
        <Option.Icon type="oClose" align="right" />
      </Option>
    )
    const avatarComponent = getByTestId("avatar")
    const iconComponent = container.querySelector('[type="oClose"]')

    // Container should render an avatar with correct styles
    expect(container).toContainElement(avatarComponent)
    expect(avatarComponent).toHaveStyle("align-self: flex-start")

    // Icon should also have the correct alignment
    expect(iconComponent).toHaveStyle("align-self: center")
    expect(asFragment()).toMatchSnapshot()
  })
})
