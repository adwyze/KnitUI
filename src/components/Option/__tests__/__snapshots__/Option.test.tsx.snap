// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option renders OptionGroup Correctly 1`] = `
<DocumentFragment>
  .c2 {
  font-size: 1.4rem;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #1a1a1a;
  padding: 0.7rem 1.4rem;
  border: 0.1rem solid transparent;
  cursor: default;
  overflow: hidden;
  background-color: #ffffff;
}

.c2:active {
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0 0 0.2rem #0066ff;
  z-index: 1;
}

.c2:hover {
  background-color: #f2f2f2;
}

.c1 {
  font-size: 1.2rem;
  line-height: 1.8rem;
  color: #808080;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c4 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: initial;
}

<div
    class="c0"
  >
    <span
      class="c1"
    >
      SOME ITEMS
    </span>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c0"
        >
          <span
            class="c4"
          >
            Option 1
          </span>
          <span
            class="c1"
          >
            Kimetsu no Yaiba
          </span>
        </div>
      </div>
      <div
        class="c5"
      />
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c0"
        >
          <span
            class="c4"
          >
            Option 2
          </span>
          <span
            class="c1"
          >
            Kimetsu no Yaiba
          </span>
        </div>
      </div>
      <div
        class="c5"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Option renders basic Option correctly 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1.4rem;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #1a1a1a;
  padding: 0.7rem 1.4rem;
  border: 0.1rem solid transparent;
  cursor: default;
  overflow: hidden;
  background-color: #ffffff;
}

.c0:active {
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0 0 0.2rem #0066ff;
  z-index: 1;
}

.c0:hover {
  background-color: #f2f2f2;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c2 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: initial;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Tanjiro
      </span>
    </div>
    <div
      class="c3"
    />
    <div
      class="c4"
    />
  </div>
</DocumentFragment>
`;

exports[`Option renders disabled option correctly 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1.4rem;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #1a1a1a;
  padding: 0.7rem 1.4rem;
  border: 0.1rem solid transparent;
  cursor: default;
  overflow: hidden;
  background-color: #ffffff;
  cursor: not-allowed;
  opacity: 0.6;
}

.c0:active {
  pointer-events: none;
}

.c4 {
  font-size: 1.2rem;
  line-height: 1.8rem;
  color: #808080;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c3 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: initial;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <span
          class="c3"
        >
          Nezuko
        </span>
        <span
          class="c4"
        >
          this is the option description
        </span>
      </div>
    </div>
    <div
      class="c5"
    />
  </div>
</DocumentFragment>
`;

exports[`Option renders icons and avatar with correct alignment  1`] = `
<DocumentFragment>
  .c3 {
  width: 24px;
  height: 24px;
  overflow: hidden;
  border-radius: 50%;
  outline: none;
  cursor: pointer;
  color: #1a1a1a;
  font-size: calc(1.4 * 24px / 2.4);
  line-height: calc(2.0 * 24px / 2.4);
  -webkit-mask-image: radial-gradient(white,black);
  mask-image: radial-gradient(white,black);
  background: #f2f2f2;
  border: 1px solid #f2f2f2;
}

.c3 img {
  width: 24px;
  height: 24px;
  margin: 0;
  -webkit-user-drag: none;
  -moz-user-drag: none;
  -ms-user-drag: none;
  user-drag: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: -webkit-transform 120ms;
  -webkit-transition: transform 120ms;
  transition: transform 120ms;
}

.c3 svg {
  margin: 0;
}

.c3 svg path {
  fill: transparent;
}

.c3:hover {
  background: #e6e6e6;
  border: 1px solid #e6e6e6;
}

.c3:hover img {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c3:hover svg path {
  fill: transparent;
}

.c3:focus,
.c3:active {
  background: #e6e6e6;
  border: 1px solid #5ab2f2;
  box-shadow: 0px 0px 2px #0066ff;
}

.c3:focus svg path,
.c3:active svg path {
  fill: transparent;
}

.c3:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c3:disabled img {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.c9 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

.c0 {
  font-size: 1.4rem;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #1a1a1a;
  padding: 0.7rem 1.4rem;
  border: 0.1rem solid transparent;
  cursor: default;
  overflow: hidden;
  background-color: #ffffff;
}

.c0:active {
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0 0 0.2rem #0066ff;
  z-index: 1;
}

.c0:hover {
  background-color: #f2f2f2;
}

.c6 {
  font-size: 1.2rem;
  line-height: 1.8rem;
  color: #808080;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c5 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: initial;
}

.c8 {
  margin-right: 0;
  margin-left: 1.5rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c8:first-child {
  margin-left: 0;
}

.c2 {
  margin-right: 1.5rem;
  margin-left: 0;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c2:first-child {
  margin-left: 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2 c3"
        data-testid="avatar"
        title="anshumankun"
      >
        A
      </button>
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          Option 1
        </span>
        <span
          class="c6"
        >
          Kimetsu no Yaiba
        </span>
      </div>
    </div>
    <div
      class="c7"
    >
      <span
        class="c8 c9"
        height="18px"
        type="oClose"
        width="18px"
      >
        <svg
          fill="#000"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Option renders selected option correctly and fires onClick 1`] = `
<DocumentFragment>
  .c6 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

.c0 {
  font-size: 1.4rem;
  line-height: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #1a1a1a;
  padding: 0.7rem 1.4rem;
  border: 0.1rem solid transparent;
  cursor: default;
  overflow: hidden;
  background-color: #f2f2f2;
}

.c0:active {
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0 0 0.2rem #0066ff;
  z-index: 1;
}

.c0:hover {
  background-color: #f2f2f2;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c3 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: initial;
}

.c5 {
  margin-right: 0;
  margin-left: 1.5rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c5:first-child {
  margin-left: 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <span
          class="c3"
        >
          Nezuko
        </span>
      </div>
    </div>
    <div
      class="c4"
    >
      <span
        class="c5 c6"
        height="18px"
        type="oCheck"
        width="18px"
      >
        <svg
          fill="#000"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.00003 16.17L4.83003 12L3.41003 13.41L9.00003 19L21 6.99997L19.59 5.58997L9.00003 16.17Z"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
